scripts/gen_packaging_stats.pl

```diff
@@ -0,0 +1,220 @@
+		packagings => 1,
+	};
+
+	my $socket_timeout_ms = 3 * 60 * 60 * 60000;    # 3 hours
+	my $products_collection = get_products_collection($socket_timeout_ms);
+
+	my $products_count = $products_collection->count_documents($query_ref);
+
+	print STDERR "$name: $products_count products\n";
+
+	my $cursor = $products_collection->query($query_ref)->sort({created_t => 1})->fields($fields_ref);
+
+	$cursor->immortal(1);
+
+	my $total = 0;
+
+	my $packagings_stats_ref = {};
+
+	# Go through all products
+	while (my $product_ref = $cursor->next) {
+		$total++;
+
+		if ($total % 1000 == 0) {
+			print STDERR "$name: $total / $products_count processed\n";
+		}
+
+		# Generate stats for all countries + en:world (products from all countries)
+		if (not defined $product_ref->{countries_tags}) {
+			$product_ref->{countries_tags} = [];
+		}
+		push @{$product_ref->{countries_tags}}, "en:world";
+
+		foreach my $country (@{$product_ref->{countries_tags}}) {
+
+			# Generate stats for all categories + all (products from all categories)
+			if (not defined $product_ref->{categories_tags}) {
+				$product_ref->{categories_tags} = [];
+			}
+			push @{$product_ref->{categories_tags}}, "all";
+
+			foreach my $category (@{$product_ref->{categories_tags}}) {
+
+				# Go through all packaging components
+				if (not defined $product_ref->{packagings}) {
+					$product_ref->{packagings} = [];
+				}
+
+				foreach my $packaging_ref (@{$product_ref->{packagings}}) {
+					my $shape = $packaging_ref->{shape} || "en:unknown";
+					my $material = $packaging_ref->{material} || "en:unknown";
+
+					deep_val($packagings_stats_ref,
+						("countries", $country, "categories", $category, "shapes", $shape, "materials", $material))
+						+= 1;
+					deep_val($packagings_stats_ref,
+						("countries", $country, "categories", $category, "shapes", "all", "materials", $material))
+						+= 1;
+					deep_val($packagings_stats_ref,
+						("countries", $country, "categories", $category, "shapes", $shape, "materials", "all"))
+						+= 1;
+					deep_val($packagings_stats_ref,
+						("countries", $country, "categories", $category, "shapes", "all", "materials", "all"))
+						+= 1;
+
+					my @shape_parents = gen_tags_hierarchy_taxonomy("en", "packaging_shapes", $shape);
+					my @material_parents = gen_tags_hierarchy_taxonomy("en", "packaging_materials", $material);
+
+					# Also add stats to parent materials
+					foreach my $material_parent (@material_parents, "all") {
+						deep_val(
+							$packagings_stats_ref,
+							(
+								"countries", $country, "categories", $category,
+								"shapes", $shape, "materials_parents", $material_parent
+							)
+						) += 1;
+						deep_val(
+							$packagings_stats_ref,
+							(
+								"countries", $country, "categories", $category,
+								"shapes", "all", "materials_parents", $material_parent
+							)
+						) += 1;
+					}
+				}
+			}
+
+		}
+
+	}
+
+	# Create directories for the output if they do not exist yet
```

Comment:
There again I feel like separating in a function would help reading, but it's ok.